### PR TITLE
Performance of get_loc on non-unique MultiIndex

### PR DESCRIPTION
**Pull request by toobaz**
- [x] tests passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The second commit is related only in the sense that the docs mentioned a ``PerformanceWarning`` due to (non-)sorting, while it was actually due to (non-)uniqueness.

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -381,6 +381,7 @@ Performance Improvements
 - :func:`Series` / :func:`DataFrame` tab completion limits to 100 values, for better performance. (:issue:`18587`)
 - Improved performance of :func:`DataFrame.median` with ``axis=1`` when bottleneck is not installed (:issue:`16468`)
 - Improved performance of :func:`MultiIndex.get_loc` for large indexes, at the cost of a reduction in performance for small ones (:issue:`18519`)
+- Improved performance of :func:`MultiIndex.get_loc` for non-unique indexes, which as a consequence does not emit a ``PerformanceWarning`` any more
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.0.txt]: can you add the issue number (this PR number if no issue).

**Review by jreback — CHANGES_REQUESTED** [review]
can you add any asv's that are needed and show the results (or if there is already sufficient coverage, can you show the results)

**Comment by jreback** [comment]
rebase when you can and pls show perf

**Comment by toobaz** [comment]
> rebase when you can and pls show perf

Unfortunately I found out that performance drops dramatically in some cases... until I'm able to fix and merge #19539.

**Comment by jreback** [comment]
ok https://github.com/pandas-dev/pandas/pull/19464#issuecomment-368243824

keep open then?

**Comment by toobaz** [comment]
> keep open then?

definitely... I think they are both significant improvements, I just need to understand what's wrong with Windows

**Comment by toobaz** [comment]
Rebased, and #19539 did improve things significantly... but results are still not what one expects from a performance fix!

```
       before           after         ratio
     [2299693d]       [cd7fd6b1]
+         125±6ms          247±1ms     1.98  multiindex_object.Duplicated.time_duplicated_loc
-      14.4±0.3μs       11.9±0.1μs     0.82  multiindex_object.GetLoc.time_string_get_loc

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

I will have to investigate more.

**Comment by jreback** [comment]
can you rebase

**Comment by jreback** [comment]
closing as stale. if you'd like to continue pls ping.

**Pull request closed without merging**